extension/algo/src/function/spanning_forest.cpp

```diff
@@ -341,15 +345,93 @@ static std::unique_ptr<TableFuncBindData> bindFunc(main::ClientContext* context,
+    return std::make_unique<SFBindData>(std::move(columns), std::move(graphEntry),
+        expression_vector{srcOutput, dstOutput, relOutput},
         std::make_unique<SFOptionalParams>(input->optionalParamsLegacy));
 }
 
+static void getLogicalPlan(Planner* planner, const BoundReadingClause& readingClause,
+    expression_vector predicates, LogicalPlan& plan) {
+    auto& call = readingClause.constCast<BoundTableFunctionCall>();
+    auto bindData = call.getBindData()->constPtrCast<GDSBindData>();
+    auto op = std::make_shared<LogicalTableFunctionCall>(call.getTableFunc(), bindData->copy());
+
+    std::vector<std::shared_ptr<LogicalOperator>> nodeMaskPlanRoots;
+    for (auto& nodeInfo : bindData->graphEntry.nodeInfos) {
+        if (nodeInfo.predicate == nullptr) {
+            continue;
+        }
+        auto& node = nodeInfo.nodeOrRel->constCast<NodeExpression>();
+        planner->getCardinliatyEstimatorUnsafe().init(node);
+        auto p = planner->getNodeSemiMaskPlan(SemiMaskTargetType::GDS_GRAPH_NODE, node,
+            nodeInfo.predicate);
+        nodeMaskPlanRoots.push_back(p.getLastOperator());
+    }
+
+    for (auto root : nodeMaskPlanRoots) {
+        op->addChild(root);
+    }
+    op->computeFactorizedSchema();
+    planner->planReadOp(std::move(op), predicates, plan);
+
+    for (auto i = 0u; i < 2; ++i) {
+        auto nodeOutput = bindData->output[i]->ptrCast<NodeExpression>();
+        KU_ASSERT(nodeOutput != nullptr);
+        planner->getCardinliatyEstimatorUnsafe().init(*nodeOutput);
+        auto scanPlan = planner->getNodePropertyScanPlan(*nodeOutput);
+        if (!scanPlan.isEmpty()) {
+            expression_vector joinConditions;
+            joinConditions.push_back(nodeOutput->getInternalID());
+            planner->appendHashJoin(joinConditions, JoinType::INNER, plan, scanPlan, plan);
+        }
+    }
+    auto relOutput = bindData->output[2]->ptrCast<RelExpression>();
+    KU_ASSERT(relOutput != nullptr);
+    auto scanPlan = LogicalPlan();
+    auto boundNode = relOutput->getSrcNode();
+    auto nbrNode = relOutput->getDstNode();
+    const auto extendDir = ExtendDirection::FWD;
+    planner->appendScanNodeTable(boundNode->getInternalID(), boundNode->getTableIDs(), {},
+        scanPlan);
+
+    auto relProperties = planner->getProperties(*relOutput);
+
+    // If the query returns a specific relationship property (e.g., RETURN ... rel.someProperty
+    // ...), getProperties() will not include the relationship's internal ID, so we need to add it
+    // manually.
+    //
+    // If the query returns the whole relationship (e.g., RETURN ... rel ...),
+    // getProperties() will include the internal ID automatically.
+    // In that case, we must avoid adding it again to relProperties.
+    bool foundInternalId = false;
+    for (const auto& property : relProperties) {
+        if (property->getDataType().getLogicalTypeID() == LogicalTypeID::INTERNAL_ID) {
```

Comment:
This is not safe because there could be other expression with type INTERNAL_ID. Use `*property == *relOutput.getInternalID()` instead.